### PR TITLE
Memcached Sample

### PR DESCRIPTION
**Pull request by praenubilus**
solution to run Memcached in mystikos. Will test Memcached CRUD operations.
dependent issues:
- syscall prlimit: issue https://github.com/deislabs/mystikos/issues/22
- syscall prctl: issue https://github.com/deislabs/mystikos/issues/21
- userid issue: issue https://github.com/deislabs/mystikos/issues/20
- epoll issue: https://github.com/deislabs/mystikos/issues/234
- Issue https://github.com/deislabs/mystikos/issues/318

### CODE REVIEW

**Comment by jxyang** [comment]
This is also blocked by https://github.com/deislabs/mystikos/issues/234

**Comment by jxyang** [comment]
#20 #21 #22 have all been addressed. I think you can pass the identity checks now by using `-u $USER`.

We still need to fix #234. @mikbras

**Comment by praenubilus** [comment]
#234 has been fixed. Now blocked by https://github.com/deislabs/mystikos/issues/318

**Comment by jxyang** [comment]
Verified memcached works with Mike's PR: https://github.com/deislabs/mystikos/pull/324

**Comment by praenubilus** [comment]
> Verified memcached works with Mike's PR: #324

Will PR #324 being merged soon? Or it is just an experiment.

**Comment by jxyang** [comment]
It's an experiment. Mike is working on a formal fix.

**Comment by jxyang** [comment]
@praenubilus #324 is merged. That should be the last hurdle to enable memcached.

File: samples/memcached/memcached_test.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+## Connect
+print(f"connecting to {ip}:{port}")
+
+try:
+    client = Client((ip, port), connect_timeout=TIMEOUT_SEC, timeout=TIMEOUT_SEC)
+
+    key, value = "some_key", "some_value"
+
+    # cmd "set", create/update key:value
```
Comment (jxyang) [inline samples/memcached/memcached_test.py:16]: Please try the CRUD operations in a loop several times.

File: samples/memcached/Makefile [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+run-host: clean all
+	docker stop memcached-myst || true
+	docker run --name memcached-myst --rm -d -p 11211:11211 memcached:1.6.9-alpine memcached -m 64
+	python3 memcached_test.py $(SERVICE_ADD) $(SERVICE_PORT)
+	docker stop memcached-myst
+
+gdb: clean all
+	@ echo -e "\n------start debugging memcached server------\n"
+	$(MYST_GDB) --args $(MYST_EXEC) $(OPTS) rootfs /usr/local/bin/memcached
```
Comment (jxyang) [inline samples/memcached/Makefile:30]: Add `-u root` as well.

File: samples/memcached/Makefile [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+SERVICE_ADD=localhost
+SERVICE_PORT=11211
+
+all: myst cpio
+
+run: 
+	# Kill the running instance of the server before exit.
+	@ echo -e "\n------starting memcached server------\n"
+	# Launch the Memcached Server
+	$(MYST_EXEC) $(OPTS) rootfs /usr/local/bin/memcached -u root start &
```
Comment (jxyang) [inline samples/memcached/Makefile:17]: Do we need `start` as a command line option?

**Comment by praenubilus** [comment]
> @praenubilus #324 is merged. That should be the last hurdle to enable memcached.

I will update some out dated settings and verify it

**Comment by paulcallen** [comment]
can you try to see if this works. if so please update with latest status/ If it works then we should work to get this checked in.

**Comment by praenubilus** [comment]
updated and tested. Ready for merge

**Comment by jxyang** [comment]
Since this one has uncovered so many bugs, can we add it under solutions or tests to be included in the pipeline?

**Comment by jxyang** [comment]
You need to install pymemcache for the client as part of the makefile.

**Comment by praenubilus** [comment]
Updated with client in container now.

**Review by jxyang — APPROVED** [review]

**Pull request merged**